{"cells": [{"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["import numpy as np\n", "import pandas as pd\n", "import os\n", "import requests\n", "from sklearn.preprocessing import OneHotEncoder, LabelEncoder"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Function to fetch CRY1 gene sequence from UCSC API"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["def get_CRY1_gene():\n", "    response = requests.get(\"https://api.genome.ucsc.edu/getData/sequence?genome=hg38;chrom=chr12;start=106991364;end=107093549\")\n", "    if response.status_code == 200:\n", "        return response.json().get(\"dna\", \"\").upper()\n", "    else:\n", "        return None"]}, {"cell_type": "markdown", "metadata": {}, "source": ["One-hot encode sequence and return a 1D array"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["def onehotencoder(fasta_sequence, max_length = 102500):\n", "    sequence_array = np.array(list(fasta_sequence))\n", "    label_encoder = LabelEncoder()\n", "    integer_encoded = label_encoder.fit_transform(sequence_array)\n", "    onehotencoder = OneHotEncoder(sparse_output=False, dtype = np.float32)\n", "    integer_encoded = integer_encoded.reshape(len(integer_encoded), 1)\n", "    onehot_sequence = onehotencoder.fit_transform(integer_encoded).astype(np.float32)\n", "    if onehot_sequence.shape[0] < max_length:\n", "        pad_size = max_length - onehot_sequence.shape[0]\n", "        padding = np.zeros((pad_size, onehot_sequence.shape[1]))\n", "        onehot_sequence = np.vstack([onehot_sequence, padding])\n", "    else:\n", "        onehot_sequence = onehot_sequence[:max_length, :] \n", "    return(onehot_sequence.flatten())\n", "# Function to add mutation to the CRY1 sequence\n", "def add_mutation(seq, start, end, refnuc, altnuc):\n", "    seq_start = start \n", "    seq_end = end\n", "    if seq[seq_start:seq_end] != refnuc:\n", "        print(f\"Mutation mismatch at positions {start}-{end}.\")\n", "        return None\n", "    else:\n", "        mutated_seq = seq[:seq_start] + altnuc + seq[seq_end:]\n", "        return mutated_seq"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Process the mutation CSV and apply mutations to the CRY1 sequence"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["def process_mutations(csv_path, cry1_seq, output_path):\n", "    df = pd.read_csv(csv_path, usecols=['chromEnd', 'ref', 'alt', 'AF', 'genes', 'variation_type', '_displayName'])\n", "    df = df[df[\"variation_type\"].str.contains(\"intron_variant\", na=False, case=False)]\n\n", "    # Initialize existing data for appending\n", "    if os.path.exists(output_path):\n", "        existing_data = np.load(output_path, allow_pickle=True)[\"arr_0\"].tolist()\n", "    else:\n", "        existing_data = []\n\n", "    # Process each row in the CSV\n", "    seq_count = 0\n", "    batch = 100\n", "    rows_save = []\n", "    max_seq_count = 100\n", "    for index, row in df.iterrows():\n", "        if seq_count >= max_seq_count:\n", "            print(\"Reached maximum sequence count.\")\n", "            break\n", "        gnomAD_ID = row[\"_displayName\"]\n", "        refnuc = str(row[\"ref\"]) if pd.notna(row[\"ref\"]) else \"\"\n", "        start = row[\"chromEnd\"] - 106991364 - len(row[\"ref\"])\n", "        end = row[\"chromEnd\"] - 106991364\n", "        altnuc = str(row[\"alt\"]) if pd.notna(row[\"alt\"]) else \"\"\n\n", "        # Add mutation to the sequence\n", "        mutated_seq = add_mutation(cry1_seq, start, end, refnuc, altnuc)\n", "        print(f\"Processed mutation {gnomAD_ID} at positions {start}-{end}.\")\n", "        if mutated_seq:\n", "            encoded_seq = onehotencoder(mutated_seq)\n", "            seq_count += 1\n", "            rows_save.append(encoded_seq)\n", "            if len(rows_save) >= batch:\n", "                existing_data.extend(rows_save)\n", "                np.savez_compressed(output_path, arr_0=np.array(existing_data))\n", "                rows_save = []\n", "    \n", "    if rows_save:\n", "        existing_data.extend(rows_save)\n", "        np.savez_compressed(output_path, arr_0=np.array(existing_data))\n", "    print(f\"Processed {seq_count} mutated sequences and saved to {output_path}.\")"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Path to CSV file with mutations and output path"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["csv_file = \"cry1realvariations (1).csv\"\n", "output_file = \"E:\\\\datasets\\\\processeddata\\\\MUTATION_DATA_TRAINING_100.npz\""]}, {"cell_type": "markdown", "metadata": {}, "source": ["Fetch CRY1 sequence and process mutations"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["cry1_seq = get_CRY1_gene()\n", "if cry1_seq:\n", "    process_mutations(csv_file, cry1_seq, output_file)\n", "else:\n", "    print(\"Failed to fetch CRY1 gene sequence.\")"]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}