# CrIMR: Integrating Machine Learning with CRY Mutation Sequencing and Repressor Gene Expression Analysis for Personalized DSPD Diagnosis and Treatment

Genetic sleep disorders, specifically Delayed Sleep Phase Syndrome (DSPS), affect 15% of the adolescent and adult population, with 60% of people with DSPD eventually having depression. Of these cases, 40-50% of them are genetically caused. Characterized by a disrupted sleep-wake cycle and circadian rhythm, this disorder is mainly caused by the Cryptochrome circadian repressor 1 (CRY1) gene, which inhibits CLOCK:BMAL 1 activity. Mutations in the gene disrupt a splice site, causing an in-frame deletion of exon 11 that modulates the period of circadian cycling, delaying sleep-wake cycles. Currently, methods for diagnosing DSPD, such as DLMOs, Actigraphy, and Bright-light therapy neglect the genetic causation of this disorder. In this research, we developed CrIMR, a LSTM-based RNN using CRY1 mutation sequencing and repressor gene expressions for personalized DSPS diagnosis and treatment. Trained off of UCSC Browser's 6000 mutated and nonmutated sequences, the RNN locates CRY1 intron mutations within the gene, and using the Hillâ€™s function with splicing scores, exact changes in repression are figured out to provide hyper-personalized treatment. For clinical implementation, the Flask-drive DSPDiagnosis was created, hosting CrIMR to provide easy access to this solution. This model highlights a 94.2% accuracy across 50 epochs of training, with an F1 score of 93.8% and 94.5% for mutated and nonmutated datasets respectively, while performing 6-30% higher in terms of accuracy compared to current diagnosis methods. CrIMR's multifaceted approach leads to the identification of mutations and repression data for accurate DSPS diagnoses and assessment of the impact it has on sleep-wake cycles.